DBus backend

donadigo opened the pull request:
This branch swaps the previous backend which was communicating with the Plug through command line and again, string values and weird parsing to determine / set actions for a DBus-based service. The service contains methods `SetKey` and `GetKey` to retrieve and set values for the logind config file. `SetKey` requires the sender to have `org.pantheon.switchboard.power.administration` ID permissions in order to write to the `logind.conf` file. There is also `Changed` signal which is invoked when `SetKey` set the value of key successfully. There is a minor `GetConfigFile` which will return the full path to the logind config file (this method isn't used in the plug itself). The service is available on demand and isn't running all the time on the system. After `ACTIVE_TIMEOUT_SECONDS` (which is 5 seconds at the moment) the service will shutdown after no activity (that is, when the server didn't receive any calls in this time).

kvdz commented on cli/Server.vala:
```diff
@@ -0,0 +1,174 @@
+ * Boston, MA  02110-1301, USA.
+ */
+
+[DBus (name = "org.freedesktop.DBus")]
```
put dbus in a interface file it us ugly above the class

kvdz commented on cli/Server.vala:
```diff
@@ -0,0 +1,174 @@
+     *
+     * Emitted when a call to set_key () succeeded
+     */
+    public signal void changed ();
```
group variables, signals and properties in the top of the file

kvdz commented on cli/Server.vala:
```diff
@@ -0,0 +1,174 @@
+     */
+    public signal void changed ();
+
+    private bool _present = false;
```
group variables, signals and properties in the top of the file

kvdz commented on cli/Server.vala:
```diff
@@ -0,0 +1,174 @@
+        }
+    }
+
+    private static Server? instance = null;
```
group variables, signals and properties in the top of the file

kvdz commented on cli/main.vala:
```diff
@@ -17,52 +17,66 @@
+    public override void activate () {
+        own_id = Bus.own_name (BusType.SYSTEM, IFACE_NAME, BusNameOwnerFlags.REPLACE,
+                    on_bus_acquired,
+                    () => {},
```
why leave bus name acquired empty?

kvdz commented on src/LidCloseActionComboBox.vala:
```diff
@@ -43,75 +44,92 @@ namespace Power {
-        private CliCommunicator.Action get_action () {
-            CliCommunicator.Action action = CliCommunicator.Action.NOT_SUPPORTED;
-            debug ("active:%d", active);
+        private Utils.Action get_action () {
```
to many early returns this is bad coding design.

kvdz commented on src/Utils.vala:
```diff
@@ -0,0 +1,78 @@
+ * Boston, MA  02110-1301, USA.
+ */
+
+namespace Power {
```
dont use utils for a class that can be separated. This can easy be a loginhelper.vala file with his own class.

donadigo commented:
@kvdz: I've addressed some of your issues in the recent commit, although I can't justify some of the things you wrote about the proposed code. I couldn't find any information about MIRCA or SEI ISO standards for code design. Could you link me to these, or explain what they state? I know there's S.O.L.I.D code principle but I don't see how my code could violate it. Can you elaborate and point out what exact issues does my code has? I'm interested in feedback. 

Regarding the changes, I've moved the interface to the Interface.vala file that's in the src/ folder, and the cli/ part of code now just compiles with this file. The callback for BusNameAcquired is left out `null` simply because there is nothing to do when the name is acquired, we expose the DBus interface only when the connection to it was successfully obtained. I'm also not sure about your comment about having to "many early returns that are bad coding design". Can you explain why that is bad? The `get_action` method doesn't do anything afterwards, so I decided to return the result immediately rather than assigning it to a temporary one and returning it at the end. If you really want that to be done, I'll push another commit for that.

kvdz commented:
here can you read about sei cert standard which keeps the misra c standard. (misra c standard is not freely available) 
https://www.securecoding.cert.org/confluence/plugins/servlet/mobile#content/view/285

the early returns is not a technical issue it is a readability issue. wat the sei cert and the misra c standard have the goals of ISO 9000  in them. like readability of the code. the problem with using early returns  is that other developers can miss the returns and hack around it. that ks why this design alway should be as straight foreword as possible. that is why one function one return is a rule in misra c. 

the se counts for a utils class. this is design wise not compliant with solid principles specially with the S of solid. simple principle. the routines in utils are not utiles related. they should have there own clear class. 

defining varaibles global should be grouped at the top of the file. this has todo with compiling c. c can not coupe correctly with declaring variables above the function. it also makes the code harder to read, because finding it is harder. 

note  sins it is opensource software and doesn't needs clasifications like KIWA or the amerikan version FDA
fully compliant tk these standards is not a hard specified.
to me it is better to read and understand when these comments are incorporated.

donadigo commented:
@kvdz can I get a re-review?
